datadog/resource_datadog_cloud_configuration_rule.go

```diff
@@ -0,0 +1,307 @@
+	complianceRuleOptions.SetRegoRule(*datadogV2.NewCloudConfigurationRegoRule(regoPolicy, resourceTypes))
+	complianceRuleOptions.SetComplexRule(isComplexRule)
+	return complianceRuleOptions
+}
+
+func getAllResourceTypes(d *schema.ResourceData) (string, []string, bool) {
+	mainResourceType := d.Get(resourceTypeField).(string)
+	relatedResourceTypes := getStringSlice(d, relatedResourceTypesField)
+
+	resourceTypes := make([]string, 0)
+	resourceTypes = append(resourceTypes, mainResourceType)
+	resourceTypes = append(resourceTypes, relatedResourceTypes...)
+
+	return mainResourceType, resourceTypes, len(resourceTypes) > 0
+}
+
+func buildRuleUpdateCases(d *schema.ResourceData, ctx context.Context) []datadogV2.SecurityMonitoringRuleCase {
+	notifications := getStringSlice(d, notificationsField)
+	severity := d.Get(severityField).(string)
+
+	ruleCase := datadogV2.NewSecurityMonitoringRuleCase()
+	ruleCase.SetStatus(datadogV2.SecurityMonitoringRuleSeverity(severity))
+	ruleCase.SetNotifications(notifications)
+
+	return []datadogV2.SecurityMonitoringRuleCase{*ruleCase}
+}
+
+func cloudConfigurationRuleReadContext(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	providerConf := meta.(*ProviderConfiguration)
+	apiInstances := providerConf.DatadogApiInstances
+	auth := providerConf.Auth
+
+	id := d.Id()
+	ruleResponse, httpResponse, err := apiInstances.GetSecurityMonitoringApiV2().GetSecurityMonitoringRule(auth, id)
+	if err != nil {
+		if httpResponse != nil && httpResponse.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+	if err := utils.CheckForUnparsed(ruleResponse); err != nil {
+		return diag.FromErr(err)
+	}
+
+	standardRuleResponse := ruleResponse.SecurityMonitoringStandardRuleResponse
+	if standardRuleResponse != nil {
+		if standardRuleResponse.GetType() == datadogV2.SECURITYMONITORINGRULETYPEREAD_CLOUD_CONFIGURATION {
+			updateResourceDataFromResponse(d, standardRuleResponse, ctx)
+		} else {
+			return diag.Errorf("Rule with id %s is not a cloud_configuration rule. This terraform resource can only manage `cloud_configuration` rules", d.Id())
+		}
+	}
+	return nil
+}
+
+func updateResourceDataFromResponse(d *schema.ResourceData, ruleResponse *datadogV2.SecurityMonitoringStandardRuleResponse, ctx context.Context) {
+	d.Set(messageField, ruleResponse.GetMessage())
+	d.Set(nameField, ruleResponse.GetName())
+	d.Set(enabledField, ruleResponse.GetIsEnabled())
+
+	resourceType := ruleResponse.GetOptions().ComplianceRuleOptions.ResourceType
+	regoRule := ruleResponse.GetOptions().ComplianceRuleOptions.GetRegoRule()
+	d.Set(resourceTypeField, resourceType)
+	d.Set(relatedResourceTypesField, getRelatedResourceTypes(*resourceType, regoRule.ResourceTypes))
+	d.Set(policyField, regoRule.Policy)
+
+	ruleCase := ruleResponse.GetCases()[0]
+	d.Set(severityField, ruleCase.Status)
+	d.Set(notificationsField, ruleCase.Notifications)
```

Comment:
For optional fields, need to check wether the field is set first. Something like:
```
	if v, ok := ruleCase.GetNotificationsOk(); ok && len(v) > 0 {
                d.Set(notificationsField, *v)
	}
```

Reply:
For `cloud_configuration` rules, cases must have exactly one item with status set, see: https://github.com/DataDog/logs-backend/blob/prod/domains/cloud-security-platform/apps/security-monitoring-api/src/main/java/com/dd/logs/api/rule_engine/validators/compliance/ComplianceRuleValidator.java#L123

Reply:
@symphony-elias `status` yes but `Notifications` is `Optional: true` as far as I can tell?

Reply:
Yes but the comment was more to handle the case when `notifications` was not in the json payload of the response. I replaced `ruleCase.Notifications` with `ruleCase.GetNotifications()` which returns an empty array if the field is not present. Empty array is the default value when not specified in the TF resource.